src/cdk-experimental/table/virtual-rows.ts

```diff
@@ -0,0 +1,292 @@
+  }
+
+  stickyHeaderRowsUpdated(update: StickyUpdate): void {
+    this._listener?.stickyHeaderRowsUpdated(update);
+  }
+}
+
+
+/**
+ * A directive that enables virtual scroll for a {@link CdkTable}.
+ */
+@Directive({
+  selector: 'cdk-table[virtualRows], table[cdk-table][virtualRows]',
+  exportAs: 'cdkVirtualRows',
+  providers: [
+    {provide: _VIEW_REPEATER_STRATEGY, useClass: _RecycleViewRepeaterStrategy},
+    // The directive cannot provide itself as the sticky positions listener because it introduces
+    // a circular dependency. Use an intermediate listener as a proxy.
+    {provide: STICKY_POSITIONING_LISTENER, useClass: PositioningListenerProxy},
+  ],
+})
+export class CdkVirtualRows<T> implements CdkVirtualScrollRepeater<T>, OnDestroy, StickyPositioningListener {
+  /** Emits when the component is destroyed. **/
+  private _destroyed = new ReplaySubject<void>(1);
+
+  /** The currently rendered range of indices. */
+  private _renderedRange: ListRange;
+
+  /** Subject that emits when a new DataSource instance is given. */
+  private _dataSourceChanges = new Subject<CdkTableDataSourceInput<T>>();
+
+  /** The original data source passed from the parent component. */
+  private _innerDataSource: CdkTableDataSourceInput<T>;
+
+  /** Emits when the virtual scroll viewport's CSS `translateY()` value changes. */
+  private _renderedContentOffsetChanged = this._viewport.renderedContentOffsetChanged.pipe(
+      map(() => this._viewport.getOffsetToRenderedContentStart()),
+      filter(offset => offset !== null) as OperatorFunction<number|null, number>,
+      distinctUntilChanged());
+
+  /** Emits when the header rows sticky state changes. */
+  private _headerRowStickyUpdates = new Subject<StickyUpdate>();
+
+  /** Emits when the footer rows sticky state changes. */
+  private _footerRowStickyUpdates = new Subject<StickyUpdate>();
+
+  /**
+   * Observable that emits the data source's complete data set. This exists to implement
+   * {@link CdkVirtualScrollRepeater}.
+   */
+  readonly dataStream: Observable<T[] | ReadonlyArray<T>> =
+      this._dataSourceChanges.asObservable()
+      .pipe(
+          switchMap(dataSource => this.getDataStream(dataSource)),
+          shareReplay(1));
+
+  /**
+   * The datasource for the table. The datasource is wrapped in another observable before passing
+   * it to the table component. This getter will return the original datasource, but it will not be
+   * the same instance received by the table.
+   */
+  @Input()
+  get dataSource(): CdkTableDataSourceInput<T> {
+    return this._innerDataSource;
+  }
+  set dataSource(dataSource: CdkTableDataSourceInput<T>) {
+    if (dataSource !== this._innerDataSource) {
+      this._innerDataSource = dataSource;
+      this._dataSourceChanges.next(dataSource);
+      // Wrap the data source in a virtualized list so it only emits a subset of rows as the user
+      // scrolls.
+      this._table.dataSource = this.virtualizeDataSource(dataSource);
```

Comment:
I was able to refactor this out with two changes:

- Added a `_TABLE_COLLECTION_VIEWER` provider that overrides the table's collection viewer observable. This is necessary because the table's current `viewChange` observable is a behavior subject that immediately emits a range spanning the entire data set. This causes the table to initially load the entire data set before it is eventually replaced by the virtual scroll viewport's range.
- Added a `_dataSourceChanges` observable to the `CdkTable` that emits when the data source changes. The virtual scroll directive uses this to update its `datastream`.